### ISSUE TITLE
1.0.0 Milestone

### ISSUE THREAD

**Issue by breerly** [opentracing/opentracing-go#112]
I'm hoping to get a sense of when we will be committing to the API of this library - can we create a 1.0.0 Milestone that tracks the remaining issues?

There are many libraries that are coming to rely on this one, and being in a pre-production release  `0.x.y` means that other libraries must pin to `>= 0.9, < 0.10` to avoid letting in breaking changes.

It would be better if we could pin to `>= 1,  < 2`, which will allow both minor and patch releases to be let in without additional releases by depending libraries.

**Comment by bensigelman** [comment]
@breerly the only thing on my list is the key-value logging story, which is actually in an RFC PR right now. https://github.com/opentracing/opentracing-go/pull/108

As for the 1.0 release, that's tracked at the language-independent OT level here: https://github.com/opentracing/opentracing.github.io/milestone/2

Note that the immutable-SpanContext change mentioned there has already happened in Go and is nearly done across the board.

**Comment by bensigelman** [comment]
Also note that semver is tough for OT, because there is both the abstract OT "spec" (which is language-independent and thus unfriendly to true semver) and the individual language implementations which can theoretically introduce breaking changes without even moving OT "versions".

I would like to come up with a clever "A, B, C" style name for the major OT releases but haven't been feeling sufficiently witty yet.

**Comment by bensigelman** [comment]
TL;DR, though: I suspect that we'll be at Go 1.x by this time next week.

**Comment by breerly** [comment]
I see, so we need the spec to hit 1.0 before this library could.

**Comment by bensigelman** [comment]
@breerly 

> I see, so we need the spec to hit 1.0 before this library could.

Not officially... e.g., opentracing-py hit 1.0 already. @yurishkuro asked about this on Gitter yesterday eve and I suggested 0.9 just because we literally have a breaking change (or at least the precursor to one) in PR right now and it seemed silly to do 1.0 given that reality.

**Comment by breerly** [comment]
@bensigelman agreed, but I'd still like to see a 1.0.0 milestone defined in this library (using Github issues) so that we can quantify what's holding us back. Is it just the breaking changes in that PR?

**Comment by yurishkuro** [comment]
strictly speaking the logging change in Go is backwards compatible, so we could've released 1.0 now and have logging change in 1.1

we do have this problem of matching client libs to the spec semver, but I think we can make life easier for consumers if we at least follow semver for individual language libs, with usual restrictions on API changes.

**Comment by breerly** [comment]
@yurishkuro that would be huge. Because opentracing-go is only in `0.9`, it leaves our library no choice but to constrain to `>= 0.9, < 0.10`, which will protect our customers from breaking changes, but also ensure we need to cut a new release of our library to let in each minor release.

I would like to see this library do something like this:

```
1.0.0 - major release, committing to no breaking changes within the 1.x.y
1.1.0 - this is a minor release, use it for introducing new functionality
1.1.1 - this is a patch release, use it to fix existing functionality
```

This would let libraries pin to `>= 1, < 2`, which will let new features and bug fixes in automatically, but requires us to promise no breaking changes within the `1.x.y` range.

**Comment by bensigelman** [comment]
@yurishkuro sorry if I wasn't clear: we should definitely do semver for the individual libs and I have no problem with that. It's just that, over time, the semvers of the various libraries will diverge (and that's ok). I guess we could do something funky and lock the major version to integer multiples of 10 every time there's a "big" OT spec change... So OT-go and OT-python and OT-java would all go to `10.x.x` if something significant happens in the OT spec itself.

**Comment by bensigelman** [comment]
also @yurishkuro although the logging change is backwards-compatible, the forthcoming change to remove the deprecated methods will not be.

**Comment by breerly** [comment]
@bensigelman I think keeping semver versions aligned between libs is probably the wrong goal.

That can be useful for marketing, but really what we need is a systematic way to communicate API stability, which can vary by library. It might be best to allow each library to follow semver in a completely decoupled way, only informed by their own API stability.

This can create some confusion, if the Open Tracing spec itself were to go `2.0`. In that case, people might wonder which client libraries supported with versions of the spec. I think that's a separate problem.

One way to handle that is forking the client libraries, essentially creating `opentracing2-go`. Yet another way is to handle both spec's transparently within the same library (allowing the user to toggle). Finally, you could create a visual matrix organizing which libraries supported which versions of the Open Tracing spec.

Realistically, these open standards are very difficult to evolve beyond 1.x.y.

**Comment by breerly** [comment]
Sounds like this solved.

**Comment by bensigelman** [comment]
Yep!
